### ISSUE TITLE
Failure to setup X11-forwarding should not be fatal

### ISSUE THREAD

**Issue by UnitedMarsupials** [ronf/asyncssh#336]
I have the X11-forwarding enabled by default in my `~.ssh/config`, but the OpenSSH client does not treat a failure to set it up as _fatal_.

Unfortunately, AsyncSSH does. Worse, there is no obvious way to disable the attempts to forward X11 at all.

OpenSSH client allows to explicitly set any options on command-line, perhaps, `asyncssh.connect()` ought to support such a list too...

Even if X11-forwarding should be enabled by default (a big if), there ought to be means to disable it...

**Comment by ronf** [comment]
When AsyncSSH is used as a client, you can enable/disable X11 forwarding on a per session basis when you call create_session() or one of the other variants of it such as open_session(), create_process(), create_subprocess(), or run(). Setting this to False explicitly should override the setting which is pulled in from ~/.ssh/config, and should hopefully fix the problem for you if you don't want to disable config file support entirely in your application.

Normally, x11_forwarding defaults to False. So, if you explicitly set it to True when opening a new session on an SSH connection, AsyncSSH treats it as a fatal error if the server tells you that it doesn't allow this.

Since you're setting it here "globally" though the config file and not on a specific session, I can potentially see an argument to make this non-fatal if it fails, particularly if OpenSSH behaves that way. However, unlike OpenSSH, AsyncSSH would have no good way to returning the warning information to you that X11 forwarding failed. It could output something through the logger, but there'd be no away to pass the failure information back programmatically to the caller when the session is opened, and so ignoring the failure bothers me somewhat.

**Comment by UnitedMarsupials** [comment]
> When AsyncSSH is used as a client, you can enable/disable X11 forwarding on a per session basis when you call create_session() or one of the other variants of it such as open_session(), create_process(), create_subprocess(), or run(). Setting this to False explicitly should override the setting which is pulled in from ~/.ssh/config

I'm following the basic examples in the documentation, which use all `asyncssh.connect()` -- how do I disable X11-forwarding there?

> if you explicitly set it to True when opening a new session on an SSH connection, AsyncSSH treats it as a fatal error

My setting it in `~/.ssh/config` follows OpenSSH's semantics: please, try to setup X11-forwarding, if you can. Interpreting it as a _requirement_ seems wrong... Especially because a Python script is far less likely to need X11-forwarding than an interactive ssh-session.

> and so ignoring the failure bothers me somewhat.

A new flag?..

**Comment by ronf** [comment]
Hmm. I sent a reply to this last night, but it doesn't seem to have made it through. Here it is again:

There are two steps when setting up a new SSH session. First, you do a connect() to connect to the remote host and authenticate yourself. Then, using that SSHClientConnection object returned by the connect(), you start one or more sessions. This could be an interactive shell to run multiple commands, an exec of a single command, starting up a subsystem like SFTP, or setting up something like port forwarding to tunnel data sent over TCP or UNIX domain sockets.

Unlike port forwarding, though, X11 forwarding cannot be established at the connection level. Instead, it is set up at the same time you open an interactive shell or exec a command, and each session you set up can independently request whether or not to do X11 forwarding. That's why in the AsyncSSH API you actually specify whether or not to do X11 forwarding when setting up the shell or exec session, and not as part of setting up a connection.

To be more specific, here's the simple client example from the docs:

```python
import asyncio, asyncssh, sys

async def run_client():
    async with asyncssh.connect('localhost') as conn:
        result = await conn.run('ls abc', check=True)
        print(result.stdout, end='')

try:
    asyncio.get_event_loop().run_until_complete(run_client())
except (OSError, asyncssh.Error) as exc:
    sys.exit('SSH connection failed: ' + str(exc))
```

In this example, you need to add "x11_forwarding=False" to the conn.run() call, like:

```python
        result = await conn.run('ls abc', check=True, x11_forwarding=False)
```

There are several different API calls to set up sessions, depending on what you need from them. See the list of functions in my previous reply where this might be added depending on which API you have selected.

I need to think a bit about the best way to handle this, and whether or not to make the soft failure only apply when it is enabled via config file rather than as an explicit argument, or whether to maybe as a third possible value to the existing x11_forwarding argument which would allow you to request it but not require it. I don't really have a good precedent for that in the API right now, though. All requests made at the session level are currently hard failures if any of those requests are rejected. The only exception is agent forwarding, but that's because it doesn't return a result, so you can't really tell whether it failed or not.

If I were to do this as a new argument value, I would probably call it something like `x11_forwarding='ignore_failure'`, and perhaps make it such that enabling it in the config file mapped to this new value, rather than mapping to `True`, to better preserve OpenSSH compatibility.

**Comment by UnitedMarsupials** [comment]
> In this example, you need to add "x11_forwarding=False" to the conn.run() call, like

_Every time_ I want to run something?! A script working fine for one system, will break -- and require editing every `run`-invocation to work against another?

Maybe, `connect` can accept a `kwargs` list of options matching those in OpenSSH config-file -- and overwriting anything found there -- which will become the default for all subsequent `run`-calls for the same connections?

```
     connection = asyncssh.connect(host, ForwardX11 = False, CheckHostIP = True, ...)
```

Sort of like OpenSSH client's own handling of multiple `-o` flags overwriting whatever is set in the config-file?

> x11_forwarding argument which would allow you to request it but not require it

Yeah, distinguish between `True` (attempt and require), `False` (don't attempt), and `None` (attempt, but don't require), for example. But it still ought to be possible to specify the default values for these _once_ per host -- or even once for the entire program-invocation.

**Comment by ronf** [comment]
> Maybe, connect can accept a kwargs list of options matching those in OpenSSH config-file -- and overwriting anything found there -- which will become the default for all subsequent run-calls for the same connections?

Keep in mind that OpenSSH config file support is relatively recent. Before that, all the config options already had names and defaults in prior AsyncSSH versions, and you could use the existing API calls to choose a non-default value. The only issue you're raising is that some settings are connection-level and others are session-level, and the option today exists in only one or the other place, not both. Also, the names are the pre-existing AsyncSSH option names, not the OpenSSH names. For instance, "x11_forwarding" vs. "ForwardX11Trusted".

I can see an argument for allowing some of these settings to set a non-default at the connection level, and that would become the new default at the session level, overriding the global default and current config file setting if it was specified at either the connection or session level. This change would affect more than just x11_forwarding, though, so I'm going to look into doing that as a separate check-in.

In the meantime, I have checked in commit 41c477b in the "develop" branch. It adds the new 'ignore_failure' option to x11_forwarding and also changes the ForwardX11Trusted config file option to behave as if 'ignore_failure' was set, rather than setting x11_forwarding=True.

> Yeah, distinguish between True (attempt and require), False (don't attempt), and None (attempt, but don't require), for example. But it still ought to be possible to specify the default values for these once per host -- or even once for the entire program-invocation.

I wouldn't want to use `None` like that. I don't think most people would think `None` should act like `True` except for ignoring the error. That's why I went with the more explicit 'ignore_failure'.

Regarding specifying defaults at the program invocation, that's what the SSHClientConnectionOptions class is for, at least for connection-level options. If I extend it to support session-level options as well, I think you'll get what you're looking for here. You can set options in that and pass that object to all future connect() calls, and it will hold a mix of config file options if you have requested those as well as overrides you request. Also, individual calls to connect() can further override whatever was set in the SSHClientConnectionOptions object.

**Comment by UnitedMarsupials** [comment]
> I wouldn't want to use None like that. I don't think most people would think `None` should act like `True`

It acts like `Maybe` in my proposal :-)  Or, in more technical terms, like "Unspecified" -- which means, use OpenSSH semantics.

Thank you for the low-latency response responses and attention to my report. Overall I'm too new to your software package to be a good judge of the options you're proposing...

**Comment by ronf** [comment]
Regarding specifying new defaults for session-level options in the connect() call, that's now available in the "develop" branch as commit f1cd67b. As part of this change, these values are also all now available in SSHClientConnectionOptions, so they could potentially be set once and used across multiple connections as well.

As noted above, passing in an option to create_session() overrides any value provided in connect(), and that overrides any value set via SSHClientConnectionOptions or in the config file. So, you can start with a general set of defaults and selectively override these defaults at multiple levels now.

Thanks for the suggestion. This was a relatively simple change, and it's definitely more flexible. It should also reduce confusion for people about where they need to go to set one of these options, and makes it so the documentation for both connection-level and session-level options appears in one place now.
